src/transformers/models/groupvit/modeling_groupvit.py

```diff
@@ -0,0 +1,1864 @@
+def contrastive_loss(logits: torch.Tensor) -> torch.Tensor:
+    return nn.functional.cross_entropy(logits, torch.arange(len(logits), device=logits.device))
+
+
+# Copied from transformers.models.clip.modeling_clip.clip_loss with clip->groupvit
+def groupvit_loss(similarity: torch.Tensor) -> torch.Tensor:
+    caption_loss = contrastive_loss(similarity)
+    image_loss = contrastive_loss(similarity.T)
+    return (caption_loss + image_loss) / 2.0
+
+
+def hard_softmax(logits: torch.Tensor, dim: int):
+    y_soft = logits.softmax(dim)
+    # Straight through.
+    index = y_soft.max(dim, keepdim=True)[1]
+    y_hard = torch.zeros_like(logits, memory_format=torch.legacy_contiguous_format).scatter_(dim, index, 1.0)
+    ret = y_hard - y_soft.detach() + y_soft
+
+    return ret
+
+
+def gumbel_softmax(logits: torch.Tensor, tau: float = 1, hard: bool = False, dim: int = -1) -> torch.Tensor:
+    # more stable https://github.com/pytorch/pytorch/issues/41663
+    gumbel_dist = torch.distributions.gumbel.Gumbel(
+        torch.tensor(0.0, device=logits.device, dtype=logits.dtype),
+        torch.tensor(1.0, device=logits.device, dtype=logits.dtype),
+    )
+    gumbels = gumbel_dist.sample(logits.shape)
+
+    gumbels = (logits + gumbels) / tau  # ~Gumbel(logits,tau)
+    y_soft = gumbels.softmax(dim)
+
+    if hard:
+        # Straight through.
+        index = y_soft.max(dim, keepdim=True)[1]
+        y_hard = torch.zeros_like(logits, memory_format=torch.legacy_contiguous_format).scatter_(dim, index, 1.0)
+        ret = y_hard - y_soft.detach() + y_soft
+    else:
+        # Reparametrization trick.
+        ret = y_soft
+    return ret
+
+
+def resize_attn_map(attentions, h, w, align_corners=False):
+    """
+
+    Args:
+        attentions: shape [B, groups, H*W]
+        h: height
+        w: width
+
+    Returns:
+        attentions: shape [B, groups, H, W]
+
+    """
+
+    scale = (h * w // attentions.shape[2]) ** 0.5
+    if h > w:
+        w_featmap = int(np.round(w / scale))
+        h_featmap = attentions.shape[2] // w_featmap
+    else:
+        h_featmap = int(np.round(h / scale))
+        w_featmap = attentions.shape[2] // h_featmap
+    assert (
+        attentions.shape[2] == h_featmap * w_featmap
+    ), f"{attentions.shape[2]} = {h_featmap} x {w_featmap}, h={h}, w={w}"
+
+    bs = attentions.shape[0]
+    groups = attentions.shape[1]  # number of group token
+    # [bs, groups, h*w, groups] -> [bs, groups, h, w]
+    attentions = attentions.reshape(bs, groups, h_featmap, w_featmap)
+    attentions = F.interpolate(attentions, size=(h, w), mode="bilinear", align_corners=align_corners)
+    return attentions
+
+
+def get_grouping_from_attentions(attentions, hw_shape, rescale=False, align_corners=False):
+    """
+    Args:
+        attentions (`tuple(torch.FloatTensor)`
+    Returns:
+        grouping: list[Tensor], attention map of shape [B, groups, H, W]
+    """
```

Comment:
Could you document all arguments of this function here?

Also, the format we usually follow for docstrings is detailed here: https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation